controllers/nstemplateset/space_roles.go

```diff
@@ -0,0 +1,136 @@
+	"encoding/json"
+	"fmt"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	applycl "github.com/codeready-toolchain/toolchain-common/pkg/client"
+	"github.com/pkg/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+)
+
+type spaceRolesManager struct {
+	*statusManager
+}
+
+// ensure ensures that the space roles for the users exist.
+// Returns `true, nil` if something was changed, `false, nil` if nothing changed, `false, err` if an error occurred
+// TODO: support update/removal/outdated
+func (r *spaceRolesManager) ensure(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
+	logger = logger.WithValues("nstemplateset_name", nsTmplSet.Name)
+
+	nss, err := fetchNamespacesByWorkspace(r.Client, nsTmplSet.Name)
+	if err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
+			"failed to list namespaces for workspace '%s'", nsTmplSet.Name)
+	}
+	logger.Info("ensuring space roles", "namespace_count", len(nss), "role_count", len(nsTmplSet.Spec.SpaceRoles))
+	changed := false
+	for _, ns := range nss {
+		// space roles to apply now
+		lastAppliedSpaceRoleObjs, err := r.getLastAppliedSpaceRoles(&ns)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve last applied space roles")
+		}
+		spaceRoleObjs, err := r.getSpaceRoles(&ns, nsTmplSet.Spec.SpaceRoles)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve space roles to apply")
+		}
+
+		c, err := deleteOutdatedObjects(logger, r.Client, lastAppliedSpaceRoleObjs, spaceRoleObjs)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusUpdateFailed, err, "failed to delete redundant objects in namespace '%s'", ns.Name)
+		}
+		changed = changed || c
+		// labels to apply on all new objects
+		var labels = map[string]string{
+			toolchainv1alpha1.ProviderLabelKey: toolchainv1alpha1.ProviderLabelValue,
+			toolchainv1alpha1.OwnerLabelKey:    nsTmplSet.GetName(), // TODO: is it needed?
+		}
+		logger.Info("creating space role objects")
+		for _, obj := range spaceRoleObjs {
+			logger.Info(fmt.Sprintf("creating/updating %s/%s", obj.GetNamespace(), obj.GetName()))
+		}
+		// create (or update existing) objects based the tier template
+		c, err = applycl.NewApplyClient(r.Client, r.Scheme).Apply(spaceRoleObjs, labels)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to provision namespace '%s' with space roles", ns.Name)
+		}
+		changed = changed || c
+		if changed {
+			// store the space roles in an annotation at the namespace level, so we know what was applied and how to deal with
+			// diffs when the space roles are changed (users added or removed, etc.)
```

Comment:
Updating the annotation only for the case when some object was changed wouldn't work for the case when:
1. objects are updated/deleted/created
2. the actual update of the annotation fails
3. this will trigger another reconcile
4. all objects are already up-to-date so nothing is changed
5. this part would be skipped and the annotation would stay outdated.

I guess that checking the content of the annotation at the very beginning would solve the problem - see my previous comment.

Reply:
I removed the `if changed` statement in 9ae349a since at this point (with https://github.com/codeready-toolchain/member-operator/commit/0034497a60a4d157b14e5ccacc0f538286313943), we should expect that eithr obsolete resources were removed or resources were created/updated.

Reply:
and if an error occurred when updating the annotation, then there will be a new reconcile loop, no resource will need to be created/updated/removed, but the annotation will be updated (hopefully for good this time)